File: ruby/triangle/triangle.rb
class Triangle
  attr_reader :x, :y, :z

  def initialize(x, y, z)
    fail TriangleError unless x > 0 && y > 0 && z > 0
    fail TriangleError unless x + y > z && y + z > x && z + x > y
    @x = x
    @y = y
    @z = z
  end

  def kind
    return :equilateral if x == y && y == z
    return :isosceles if x == y || y == z || x == z
    :scalene
  end
end

class TriangleError < StandardError; end

File: erlang/triangle/triangle.erl
-module(triangle).
-export([kind/3]).

kind(A, B, C) when A =< 0 orelse 
                   B =< 0 orelse 
                   C =< 0 ->
    {error, "all side lengths must be positive"};
kind(A, B, C) when A + B =< C orelse 
                   A + C =< B orelse 
                   B + C =< A->
    {error, "side lengths violate triangle inequality"};

kind(_A, _A, _A) -> equilateral;

kind(_A, _B, _B) -> isosceles;
kind(_B, _B, _A) -> isosceles;
kind(_B, _A, _B) -> isosceles;

kind(_, _, _) -> scalene.
